Fix onEnter patterns. Code and Separator blocks.

jorgefilipecosta opened the pull request:
The replace was happening on keyUp and sometimes that event did not fire. So now we do the replace right away if a pattern matches, and we cancel the event processing so no errors happen.
Sometimes the code regex did not match because of white spaces, related with the way browser handle an enter after '`' char so an optional whitespace char was added to the regex.

Because of browser behavior we now also accept \`\`\`\` followed by enter as a way to convert to the code block. Chrome automatically adds another \` if enter is pressed one time after \` so this allows the user to write \`\`\` and then press enter two times (first inserts the \`second one converts the block) to create the code block. This chrome behavior for entering is not specific to Gutenberg even in Github comment boxes it happens. Enter events did not fir when enter is inserting \`. In safari enter after \`creates a new line,  pressing two times after \`\`\`also creates a new line.

I think users using \`\`\` are aware of the strange browser use of enter and will probably press enter again as this is something, not Gutenberg specific so they are already probably aware of the behavior of enter after \`. Another alternative we have is handling '\`' keypress event so we can transform on third \` keypress without enter. The addition of this handles will make the code more complex that's why I preferred this approach.

Fixes: 
https://github.com/WordPress/gutenberg/issues/4290

## How Has This Been Tested?
Add a new paragraph write --- in it press enter and see it is transformed right away to separator block. In master, it wasn't but if after press key up to go up the transformation happened it was only transformed on the key up of the arrow not the key up of the enter.
In chrome add `` then press enter chrome automatically adds another ` now press enter and the transformation should happen. Safari creates a new line when pressing enter after `so if we insert ``` 
then press enter we get a new line if enter is pressed again the transform happens.

## Screenshots (jpeg or gifs if applicable):
Before:
![mar-23-2018 15-16-38](https://user-images.githubusercontent.com/11271197/37837572-756b618e-2ead-11e8-8e0d-93f651f90921.gif)
![mar-23-2018 15-15-39](https://user-images.githubusercontent.com/11271197/37837605-8a5d9800-2ead-11e8-8e58-bce8e1c32d03.gif)


After:
![mar-23-2018 15-00-47](https://user-images.githubusercontent.com/11271197/37837632-9c00bab0-2ead-11e8-83d1-d60adf1b1dbe.gif)
![mar-23-2018 15-10-32](https://user-images.githubusercontent.com/11271197/37837640-a1017a0e-2ead-11e8-87b7-9cadc05f822d.gif)

jorgefilipecosta committed:
Fix onEnter patterns.

The replace was happening on keyUp and sometimes that event did not fired. So now we do the replace right away if a pattern matches, and we cancel the event processing so no erros happen.
Sometimes the regex for code did not match because of white spaces, related with the way browser handle an enter after '`' char so an optional white space char was added to the regex.

aduth commented:
>Chrome automatically adds another \` if enter is pressed one time after \`

I am not familiar with this behavior, and don't encounter it myself. Why does this happen?

aduth commented on blocks/rich-text/patterns.js:
```diff
@@ -225,7 +225,8 @@ export default function( editor ) {
-
-		editor.once( 'keyup', () => onReplace( [ block ] ) );
+		onReplace( [ block ] );
+		event.preventDefault();
```
In general I'm finding `preventDefault` and `stopImmediatePropagation` to be code smells. Here their use appears legitimate, but I'd like to get in a habit of documenting their use, not unlike our "Disable reasons" for disabling ESLint rules. Here, particularly that we don't want the additional newline (`preventDefault`) and to stop TinyMCE's own processing of `keydown` which conflicts with the block replacement (since the instance is destroyed).

jorgefilipecosta commented:
> I am not familiar with this behavior, and don't encounter it myself. Why does this happen?

Interesting, so on your tests pressing ``` and then enter works as expected? And creates the code block?
Not certain about, the reason, but given that you are not facing it maybe it's because of a configuration in the S.O/keyboard, I did not find a specific reason yet. For me it happens everywhere even in this comment box I just tried.

aduth commented:
>Interesting, so on your tests pressing ``` and then enter works as expected? And creates the code block?

Correct. Reverting the ~commit for~ changes to the enter pattern for the code block, it works as expected.

jorgefilipecosta commented:
> Correct. Reverting the commit for changes to the enter pattern for the code block, it works as expected.

Thank you for confirming. I reverted the changes to the block and added the comments.

Regarding the problem of when pressing enter after backticks inserting another backtick, it is only happening in chrome, on the internet, there are persons telling other issues with backticks like they don't appear until another key is pressed, but not this specific issue, it looks the behaviour of backticks defers depending on the locale. So I think we should keep the regex as it was.

aduth commented on blocks/rich-text/patterns.js:
```diff
@@ -225,7 +225,11 @@ export default function( editor ) {
+		onReplace( [ block ] );
 
-		editor.once( 'keyup', () => onReplace( [ block ] ) );
+		// WE call preventDefault and stopImmediatePropagation in order to prevent TinyMCE from creating additional newlines.
```
Nit: Typo "WE" -> "We"

aduth replied:
Nit: `preventDefault` has nothing to do with TinyMCE. In fact, for the example from the original comment, the newline we're preventing is in a plain textarea rendered by the code block.

jorgefilipecosta replied:
Thank you for the feedback, it was applied.
